src/rpc/core_rpc_server.cpp

```diff
@@ -2735,52 +2782,60 @@ namespace cryptonote
+  {
+    std::vector<service_nodes::service_node_pubkey_info> sn_infos = m_core.get_service_node_list_state({});
 
-      entry.contributors.reserve(pubkey_info.info.contributors.size());
+    if (req.limit != 0) {
 
-      using namespace service_nodes;
-      for (service_node_info::contributor_t const &contributor : pubkey_info.info.contributors)
-      {
-        COMMAND_RPC_GET_SERVICE_NODES::response::contributor new_contributor = {};
-        new_contributor.amount   = contributor.amount;
-        new_contributor.reserved = contributor.reserved;
-        new_contributor.address  = cryptonote::get_account_address_as_str(m_core.get_nettype(), false/*is_subaddress*/, contributor.address);
+      const auto limit = std::min(sn_infos.size(), static_cast<size_t>(req.limit));
 
-        new_contributor.locked_contributions.reserve(contributor.locked_contributions.size());
-        for (service_node_info::contribution_t const &src : contributor.locked_contributions)
-        {
-          COMMAND_RPC_GET_SERVICE_NODES::response::contribution dest = {};
-          dest.amount                                                = src.amount;
-          dest.key_image                                             = string_tools::pod_to_hex(src.key_image);
-          dest.key_image_pub_key                                     = string_tools::pod_to_hex(src.key_image_pub_key);
-          new_contributor.locked_contributions.push_back(dest);
-        }
+      std::random_device rd;
+      std::mt19937 mt(rd());
 
-        entry.contributors.push_back(new_contributor);
-      }
+      std::shuffle(sn_infos.begin(), sn_infos.end(), mt);
+
+      sn_infos.resize(limit);
+    }
+
+    res.service_node_states.reserve(sn_infos.size());
+
+    for (auto &pubkey_info : sn_infos) {
+      COMMAND_RPC_GET_N_SERVICE_NODES::response::entry entry = {res};
 
-      entry.total_contributed             = pubkey_info.info.total_contributed;
-      entry.total_reserved                = pubkey_info.info.total_reserved;
-      entry.staking_requirement           = pubkey_info.info.staking_requirement;
-      entry.portions_for_operator         = pubkey_info.info.portions_for_operator;
-      entry.operator_address              = cryptonote::get_account_address_as_str(m_core.get_nettype(), false/*is_subaddress*/, pubkey_info.info.operator_address);
-      entry.swarm_id                      = pubkey_info.info.swarm_id;
+      fill_sn_response_entry(entry, pubkey_info);
 
       res.service_node_states.push_back(entry);
     }
 
+    res.status = CORE_RPC_STATUS_OK;
+    res.height = m_core.get_current_blockchain_height();
+    res.block_hash = string_tools::pod_to_hex(m_core.get_block_id_by_height(res.height - 1));
+
+    // check if any field is set
+    char unset_bytes[sizeof(req.fields)];
+    memset(unset_bytes, 0, sizeof(req.fields));
+
+    bool some_set = memcmp(unset_bytes, reinterpret_cast<const void*>(&req.fields), sizeof(req.fields));
```

Comment:
Hmm. On second thought, I think this might not be correct due to padding, technically I suppose the struct is all bools, so you wouldn't expect padding between elements except at the end. 

But unless we're sure the request field is initialized using memset when we receive it, then there could be garbage at the tail end of the structure in the padding and trigger this. And I'm guessing that's not using memset seeing as the other calls use epee::struct_init which uses aggregate initialisation and that doesn't initialise padding to 0.

I can imagine it working if you pragma pack the structure to disallow padding. unset_bytes will have to be the same type to ensure the compiler doesn't do anything unexpected, and memset unset_bytes to 0.

Reply:
I changed the macro instead (keeping the old one intact) to remember if a non-default value was found in deserialisation and serialise based on that.